File: app/src/main/java/ru/voronezhtsev/weatherapp/domain/WeatherInteractor.java
package ru.voronezhtsev.weatherapp.domain;

import io.reactivex.Single;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;
import ru.voronezhtsev.weatherapp.data.repositories.WeatherRepository;
import ru.voronezhtsev.weatherapp.models.domain.WeatherInfo;


public class WeatherInteractor {

    private WeatherRepository mWeatherRepository;
    private ILocationRepository mLocationRepository;
    private IWeatherLocalRepository mWeatherLocalRepository;

    public WeatherInteractor(WeatherRepository weatherRepository, ILocationRepository locationRepository,
                             IWeatherLocalRepository weatherLocalRepository) {
        mWeatherRepository = weatherRepository;
        mLocationRepository = locationRepository;
        mWeatherLocalRepository = weatherLocalRepository;
    }

    public Single<WeatherInfo> getWeather() {
        return mLocationRepository.getLocation()
                .flatMap(location -> mWeatherRepository.getWeather(location).subscribeOn(Schedulers.io())
                        .map(weatherInfo -> {
                            mWeatherLocalRepository.save(weatherInfo).subscribeOn(Schedulers.io())
                                    .observeOn(AndroidSchedulers.mainThread())
                                    .subscribe(() -> {
                                    }, throwable -> {
                                    });
                            return weatherInfo;
                        }));
    }
}

File: app/src/main/java/ru/voronezhtsev/weatherapp/domain/WeatherInteractor.kt
package ru.voronezhtsev.weatherapp.domain

import io.reactivex.Single
import ru.voronezhtsev.weatherapp.models.domain.Weather

class WeatherInteractor(private val weatherRepository: WeatherRepository,
                        private val forecastsRepository: ForecastsRepository) {

    /**
     * Метод загружает погоду по конкретному ИД города из сети, сохраняет данные в БД и возвращает рез-т
     */
    fun loadFromNetwork(cityId: Long): Single<List<Weather>> {
        return forecastsRepository.loadFromNetwork(cityId)
                .flatMap { weatherRepository.loadFromNetwork(cityId) }
    }

    val weatherList: Single<List<Weather>>
        get() = weatherRepository.loadFromDb()

}
